### PR TITLE
Logic App Integration Account Control Numbers cmdlets

### PR DESCRIPTION
**Pull request by daviburg**
## Description
New cmdlets for Logic App Integration Account Control Numbers:
Get, Set, List generated control numbers, Get, Remove received control numbers.

Building atop https://github.com/Azure/azure-sdk-for-net/pull/2858

Underlying sessions entities are intentionally not exposed to cmdlet user. Cmdlet user will know high level business concept of control numbers. Sessions is Microsoft's B2B connector specific underlying storage mechanism for this and other states.
Full CRUD is intentionally not exposed to cmdlet user. Select commands above are exposed for the specific purpose of a service operator to fix up current state in the event of a disaster recovery to a back-up site.

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [X] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md).**
- [X] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/dev/documentation/platyPSHelp-documentation.md).**
- [X] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#updating-the-change-log).**
For the sake of completeness, will add an additional commit with change log update to this PR.

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#general-guidelines)
- [X] Title of the pull request is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [X] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/dev/documentation/breaking-changes.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#testing-guidelines)
- [X] Pull request includes test coverage for the included changes.
- [X] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [X] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://gist.github.com/markcowl/338e16fe5c8bbf195aff9f8af0db585d#what-is-the-change).
- [X] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.
Remove cmdlet produces no output and a `PassThru` parameter would not work for it either.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [X] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [X] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [X] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

**Comment by cormacpayne** [comment]
@azuresdkci add to whitelist

**Comment by markcowl** [comment]
@davidburg this is failing because it can't find your package.  Have you published the package yet?

**Comment by shahabhijeet** [comment]
@azuresdkci test this please

File: src/ResourceManager/LogicApp/Commands.LogicApp.Test/ScenarioTests/IntegrationAccountGeneratedIcnTests.ps1 [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ----------------------------------------------------------------------------------
+
+<#
+.SYNOPSIS
+Returns true if current mode is record
+#>
+function IsRecordMode()
```
Comment (markcowl) [inline src/ResourceManager/LogicApp/Commands.LogicApp.Test/ScenarioTests/IntegrationAccountGeneratedIcnTests.ps1]: This is already availble from TestSupport.RunningMocked.  Duplicating this is  maintenance problem going forward.

File: src/ResourceManager/LogicApp/Commands.LogicApp/Cmdlets/IntegrationAccount/GetAzureIntegrationAccountAgreementCommand.cs [hunk 2]
```diff
@@ -21,22 +21,32 @@ namespace Microsoft.Azure.Commands.LogicApp.Cmdlets
     /// <summary>
     /// Gets the integration account agreement by name 
     /// </summary>
-    [Cmdlet(VerbsCommon.Get, "AzureRmIntegrationAccountAgreement"), OutputType(typeof (object))]
+    [Cmdlet(VerbsCommon.Get, "AzureRmIntegrationAccountAgreement")]
+    [OutputType(typeof(object))]
```
Comment (markcowl) [inline src/ResourceManager/LogicApp/Commands.LogicApp/Cmdlets/IntegrationAccount/GetAzureIntegrationAccountAgreementCommand.cs]: This is essentially useless, please be more specific about the output type

File: src/ResourceManager/LogicApp/Commands.LogicApp/Cmdlets/IntegrationAccount/GetAzureIntegrationAccountGeneratedIcnCommand.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+{
+    using System.Collections.Generic;
+    using System.Management.Automation;
+    using Microsoft.Azure.Commands.LogicApp.Utilities;
+
+    /// <summary>
+    /// Gets the integration account generated interchange control number by agreement name
+    /// </summary>
+    [Cmdlet(VerbsCommon.Get, "AzureRmIntegrationAccountGeneratedIcn")]
+    [OutputType(typeof(IntegrationAccountClient.IntegrationAccountControlNumber), typeof(IList<IntegrationAccountClient.QualifiedIntegrationAccountControlNumber>))]
```
Comment (markcowl) [inline src/ResourceManager/LogicApp/Commands.LogicApp/Cmdlets/IntegrationAccount/GetAzureIntegrationAccountGeneratedIcnCommand.cs]: Why areen't these simple public types?

File: src/ResourceManager/LogicApp/Commands.LogicApp/Cmdlets/IntegrationAccount/GetAzureIntegrationAccountReceivedIcnCommand.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+namespace Microsoft.Azure.Commands.LogicApp.Cmdlets
+{
+    using System.Management.Automation;
+    using Microsoft.Azure.Commands.LogicApp.Utilities;
+
+    /// <summary>
+    /// Gets the integration account received interchange control number by agreement name and control number value.
+    /// </summary>
+    [Cmdlet(VerbsCommon.Get, "AzureRmIntegrationAccountReceivedIcn")]
+    [OutputType(typeof(IntegrationAccountClient.IntegrationAccountControlNumber))]
```
Comment (markcowl) [inline src/ResourceManager/LogicApp/Commands.LogicApp/Cmdlets/IntegrationAccount/GetAzureIntegrationAccountReceivedIcnCommand.cs]: Why is this not a simple, public type?

File: src/ResourceManager/LogicApp/Commands.LogicApp/Cmdlets/IntegrationAccount/UpdateAzureIntegrationAccountGeneratedIcnCommand.cs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    using System;
+    using System.Globalization;
+    using System.Management.Automation;
+    using Microsoft.Azure.Commands.LogicApp.Utilities;
+
+    /// <summary>
+    /// Updates the integration account generated interchange control number.
+    /// </summary>
+    [Cmdlet(VerbsCommon.Set, "AzureRmIntegrationAccountGeneratedIcn", SupportsShouldProcess = true)]
+    [OutputType(typeof(IntegrationAccountClient.IntegrationAccountControlNumber))]
```
Comment (markcowl) [inline src/ResourceManager/LogicApp/Commands.LogicApp/Cmdlets/IntegrationAccount/UpdateAzureIntegrationAccountGeneratedIcnCommand.cs]: simple type

File: src/ResourceManager/LogicApp/Commands.LogicApp/Commands.LogicApp.csproj [hunk 6]
```diff
@@ -193,6 +200,10 @@
       <Project>{3819d8a7-c62c-4c47-8ddd-0332d9ce1252}</Project>
       <Name>Commands.ResourceManager.Common</Name>
     </ProjectReference>
+    <ProjectReference Include="..\..\Resources\Commands.ResourceManager\Cmdlets\Commands.Resources.Rest.csproj">
```
Comment (markcowl) [inline src/ResourceManager/LogicApp/Commands.LogicApp/Commands.LogicApp.csproj]: This is not allowed.  You cannot have a dependency on another cmdlets project

File: src/ResourceManager/LogicApp/Commands.LogicApp/Utilities/IntegrationAccountControlNumber.cs [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    using System;
+    using System.Management.Automation;
+    using Newtonsoft.Json;
+    using Newtonsoft.Json.Converters;
+    using Newtonsoft.Json.Linq;
+
+    /// <summary>
+    /// LogicApp client partial class for integration account control number types.
+    /// </summary>
+    public partial class IntegrationAccountClient
```
Comment (markcowl) [inline src/ResourceManager/LogicApp/Commands.LogicApp/Utilities/IntegrationAccountControlNumber.cs]: These look like generated types, why are these here and not in your SDK?

Comment (daviburg) [inline src/ResourceManager/LogicApp/Commands.LogicApp/Utilities/IntegrationAccountControlNumber.cs]: Because the types are client-side only. The swagger and SDK specify what the server is aware of, which here is just an obscure/opaque JToken object. The clients store state specific to themselves within the session. Different client-defined types can and are stored in different sessions.

Comment (markcowl) [inline src/ResourceManager/LogicApp/Commands.LogicApp/Utilities/IntegrationAccountControlNumber.cs]: If these are exchanged with the service, you should consider adding these types to your API definition.  This is not a blocker for the PR.

File: src/ResourceManager/LogicApp/Commands.LogicApp/Cmdlets/IntegrationAccount/UpdateAzureIntegrationAccountGeneratedIcnCommand.cs (hunk 5 above)
Comment (markcowl) [inline src/ResourceManager/LogicApp/Commands.LogicApp/Cmdlets/IntegrationAccount/UpdateAzureIntegrationAccountGeneratedIcnCommand.cs]: In other words - you need to provide a simple type definition, not a type defined within another type.

File: src/ResourceManager/LogicApp/Commands.LogicApp.Test/ScenarioTests/IntegrationAccountGeneratedIcnTests.ps1 (hunk 1 above)
Comment (daviburg) [inline src/ResourceManager/LogicApp/Commands.LogicApp.Test/ScenarioTests/IntegrationAccountGeneratedIcnTests.ps1]: So if (-not [Microsoft.WindowsAzure.Commands.Utilities.Common.TestMockSupport]::RunningMocked) gives me the wrong result when running in Record mode, i.e. return trues which cause setup code not to run and the test to fail. I will keep the approach I mimicked from existing test code which works reliably.

File: src/ResourceManager/LogicApp/Commands.LogicApp/Cmdlets/IntegrationAccount/GetAzureIntegrationAccountAgreementCommand.cs (hunk 2 above)
Comment (daviburg) [inline src/ResourceManager/LogicApp/Commands.LogicApp/Cmdlets/IntegrationAccount/GetAzureIntegrationAccountAgreementCommand.cs]: Fixing

File: src/ResourceManager/LogicApp/Commands.LogicApp/Cmdlets/IntegrationAccount/GetAzureIntegrationAccountGeneratedIcnCommand.cs (hunk 3 above)
Comment (daviburg) [inline src/ResourceManager/LogicApp/Commands.LogicApp/Cmdlets/IntegrationAccount/GetAzureIntegrationAccountGeneratedIcnCommand.cs]: Fixing

File: src/ResourceManager/LogicApp/Commands.LogicApp/Commands.LogicApp.csproj (hunk 6 above)
Comment (daviburg) [inline src/ResourceManager/LogicApp/Commands.LogicApp/Commands.LogicApp.csproj]: Removing

File: src/ResourceManager/LogicApp/Commands.LogicApp/Cmdlets/IntegrationAccount/GetAzureIntegrationAccountAgreementCommand.cs (hunk 2 above)
Comment (daviburg) [inline src/ResourceManager/LogicApp/Commands.LogicApp/Cmdlets/IntegrationAccount/GetAzureIntegrationAccountAgreementCommand.cs]: This change resulted in a breaking change build error. I think that as object return type provided no value this breaking change is unlikely to actually break any client. Is there a way to acknowledgement the breaking change as acceptable?

00:48:41   "D:\workspace\powershell\src\Package\Debug\ResourceManager\AzureResourceManager\AzureRM.LogicApp\Microsoft.Azure.Commands.LogicApp.dll","Microsoft.Azure.Commands.LogicApp.Cmdlets.GetAzureIntegrationAccountAgreementCommand","Get-AzureRmIntegrationAccountAgreement","0","1020","The cmdlet 'Get-AzureRmIntegrationAccountAgreement' no longer has output type 'Object'.","Make cmdlet 'Get-AzureRmIntegrationAccountAgreement' return type 'Object'."

**Comment by markcowl** [comment]
@daviburg You have two issues causing static analysis failures:

- Cmdlets with nouns that are plural, please fix this
-  The commands that now have a better return type than object - please add exceptiosn for these, our tool is incorrectly flagging these as breakign changes when they are not.  We have an issue to fix the tool, but adding the escepotions for now should fix the issue.

**Comment by daviburg** [comment]
I added the exceptions for the return type. I did not find in the Jenkins build artifacts or console output where is the "Cmdlets with nouns that are plural" issue.

**Comment by markcowl** [comment]
on demand run here:http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1432/ 

LGTM once the on demand run passes
